hmp/trialdata.py

```diff
@@ -0,0 +1,118 @@
+import gc
+import itertools
+import multiprocessing as mp
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from functools import cached_property
+from itertools import cycle, product
+from typing import Any
+from warnings import resetwarnings, warn
+
+import matplotlib.pyplot as plt
+import numpy as np
+import xarray as xr
+from pandas import MultiIndex
+from scipy.signal import correlate
+from scipy.stats import norm as norm_pval
+
+
+@dataclass
+class TrialData():
+    named_durations: Any
+    coords: Any
+    starts: Any
+    ends: Any
+    n_trials: int
+    n_samples: int
+    n_dims: int
+    cross_corr: Any
+    trial_coords: Any
+
+    @classmethod
+    def from_standard_data(cls, data, template):
+
+        # compute sequence durations based on number of samples
+        durations = (
+            data.unstack()
+            .sel(component=0)
+            .rename({"epochs": "trials"})
+            .stack(trial_x_participant=["participant", "trials"])
+            .dropna(dim="trial_x_participant", how="all")
+            .groupby("trial_x_participant")
+            .count(dim="samples")
+            .cumsum()
+            .squeeze()
+        )
+
+        if durations.trial_x_participant.count() > 1:
+            dur_dropped_na = durations.dropna("trial_x_participant")
+            starts = np.roll(dur_dropped_na.data, 1)
+            starts[0] = 0
+            ends = dur_dropped_na.data - 1
+            named_durations = durations.dropna("trial_x_participant") - durations.dropna(
+                "trial_x_participant"
+            ).shift(trial_x_participant=1, fill_value=0)
+            coords = durations.reset_index("trial_x_participant").coords
+        else:
+            dur_dropped_na = durations
+            starts = np.array([0])
+            ends = np.array([dur_dropped_na.data - 1])
+            named_durations = durations
+            coords = durations.coords
+
+        n_trials = durations.trial_x_participant.count().values
+        n_samples, n_dims = np.shape(data.data.T)
+        cross_corr = cross_correlation(data.data.T, n_trials, n_dims, starts, ends, template)  # Equation 1 in 2024 paper
+        trial_coords = (
+            data.unstack()
+            .sel(component=0, samples=0)
+            .rename({"epochs": "trials"})
+            .stack(trial_x_participant=["participant", "trials"])
+            .dropna(dim="trial_x_participant", how="all")
+            .coords
+        )
+        return cls(named_durations=named_durations, coords=coords, starts=starts, ends=ends,
+                   n_trials=n_trials, n_samples=n_samples, n_dims=n_dims, cross_corr=cross_corr,
+                   trial_coords=trial_coords)
+
+    @cached_property
+    def max_duration(self):
+        return self.durations.max()
+
+    @cached_property
+    def mean_duration(self):
+        return self.durations.mean()
+
+    @cached_property
+    def durations(self):
+        return self.ends - self.starts + 1
```

Comment:
Isn't this slowing things (slightly) down? That data shouldn't be too heavy so I would say we should compute it once and then just refer to that data (e.g. see the repeated calls in fixedn.estim_probs)

Reply:
Same for the mean_durations and max_durations

Reply:
Cached property computes it once and caches it, so it will only be computed once (when it is first needed).

Reply:
Oh OK perfect